### ISSUE TITLE
Errors in installing ConstraintKinds 1.1.0.0 with cabal 1.20.0.1 and ghc 7.8.2

### ISSUE THREAD

**Issue by mindbound** [mikeizbicki/ConstraintKinds#6]
The full log:

```
$ cabal install constraintkinds
Resolving dependencies...
Configuring ConstraintKinds-1.1.0.0...
Building ConstraintKinds-1.1.0.0...
Preprocessing library ConstraintKinds-1.1.0.0...
[1 of 9] Compiling Control.ConstraintKinds.Partitionable ( src/Control/ConstraintKinds/Partitionable.hs, dist/build/Control/ConstraintKinds/Partitionable.o )

src/Control/ConstraintKinds/Partitionable.hs:42:30:
    Could not deduce (RealFrac s1) arising from a use of ‘ceiling’
    from the context (PartitionableConstraint V.Vector a)
      bound by the type signature for
                 partition :: (PartitionableConstraint V.Vector a) =>
                              Int -> V.Vector a -> [V.Vector a]
      at src/Control/ConstraintKinds/Partitionable.hs:33:5-13
    The type variable ‘s1’ is ambiguous
    Note: there are several potential instances:
      instance RealFrac Double -- Defined in ‘GHC.Float’
      instance RealFrac Float -- Defined in ‘GHC.Float’
      instance Integral a => RealFrac (GHC.Real.Ratio a)
        -- Defined in ‘GHC.Real’
    In the expression: ceiling
    In the expression:
      ceiling $ (fromIntegral $ V.length vec) / (fromIntegral n)
    In an equation for ‘lenmax’:
        lenmax = ceiling $ (fromIntegral $ V.length vec) / (fromIntegral n)

src/Control/ConstraintKinds/Partitionable.hs:42:41:
    Could not deduce (Num s1) arising from a use of ‘fromIntegral’
    from the context (PartitionableConstraint V.Vector a)
      bound by the type signature for
                 partition :: (PartitionableConstraint V.Vector a) =>
                              Int -> V.Vector a -> [V.Vector a]
      at src/Control/ConstraintKinds/Partitionable.hs:33:5-13
    The type variable ‘s1’ is ambiguous
    Note: there are several potential instances:
      instance Num Double -- Defined in ‘GHC.Float’
      instance Num Float -- Defined in ‘GHC.Float’
      instance Integral a => Num (GHC.Real.Ratio a)
        -- Defined in ‘GHC.Real’
      ...plus five others
    In the expression: fromIntegral
    In the first argument of ‘(/)’, namely
      ‘(fromIntegral $ V.length vec)’
    In the second argument of ‘($)’, namely
      ‘(fromIntegral $ V.length vec) / (fromIntegral n)’

src/Control/ConstraintKinds/Partitionable.hs:42:70:
    Could not deduce (Fractional s1) arising from a use of ‘/’
    from the context (PartitionableConstraint V.Vector a)
      bound by the type signature for
                 partition :: (PartitionableConstraint V.Vector a) =>
                              Int -> V.Vector a -> [V.Vector a]
      at src/Control/ConstraintKinds/Partitionable.hs:33:5-13
    The type variable ‘s1’ is ambiguous
    Note: there are several potential instances:
      instance Fractional Double -- Defined in ‘GHC.Float’
      instance Fractional Float -- Defined in ‘GHC.Float’
      instance Integral a => Fractional (GHC.Real.Ratio a)
        -- Defined in ‘GHC.Real’
    In the second argument of ‘($)’, namely
      ‘(fromIntegral $ V.length vec) / (fromIntegral n)’
    In the expression:
      ceiling $ (fromIntegral $ V.length vec) / (fromIntegral n)
    In an equation for ‘lenmax’:
        lenmax = ceiling $ (fromIntegral $ V.length vec) / (fromIntegral n)

src/Control/ConstraintKinds/Partitionable.hs:55:30:
    Could not deduce (RealFrac s0) arising from a use of ‘ceiling’
    from the context (PartitionableConstraint VU.Vector a)
      bound by the type signature for
                 partition :: (PartitionableConstraint VU.Vector a) =>
                              Int -> VU.Vector a -> [VU.Vector a]
      at src/Control/ConstraintKinds/Partitionable.hs:46:5-13
    The type variable ‘s0’ is ambiguous
    Note: there are several potential instances:
      instance RealFrac Double -- Defined in ‘GHC.Float’
      instance RealFrac Float -- Defined in ‘GHC.Float’
      instance Integral a => RealFrac (GHC.Real.Ratio a)
        -- Defined in ‘GHC.Real’
    In the expression: ceiling
    In the expression:
      ceiling $ (fromIntegral $ VU.length vec) / (fromIntegral n)
    In an equation for ‘lenmax’:
        lenmax
          = ceiling $ (fromIntegral $ VU.length vec) / (fromIntegral n)

src/Control/ConstraintKinds/Partitionable.hs:55:41:
    Could not deduce (Num s0) arising from a use of ‘fromIntegral’
    from the context (PartitionableConstraint VU.Vector a)
      bound by the type signature for
                 partition :: (PartitionableConstraint VU.Vector a) =>
                              Int -> VU.Vector a -> [VU.Vector a]
      at src/Control/ConstraintKinds/Partitionable.hs:46:5-13
    The type variable ‘s0’ is ambiguous
    Note: there are several potential instances:
      instance Num Double -- Defined in ‘GHC.Float’
      instance Num Float -- Defined in ‘GHC.Float’
      instance Integral a => Num (GHC.Real.Ratio a)
        -- Defined in ‘GHC.Real’
      ...plus five others
    In the expression: fromIntegral
    In the first argument of ‘(/)’, namely
      ‘(fromIntegral $ VU.length vec)’
    In the second argument of ‘($)’, namely
      ‘(fromIntegral $ VU.length vec) / (fromIntegral n)’

src/Control/ConstraintKinds/Partitionable.hs:55:71:
    Could not deduce (Fractional s0) arising from a use of ‘/’
    from the context (PartitionableConstraint VU.Vector a)
      bound by the type signature for
                 partition :: (PartitionableConstraint VU.Vector a) =>
                              Int -> VU.Vector a -> [VU.Vector a]
      at src/Control/ConstraintKinds/Partitionable.hs:46:5-13
    The type variable ‘s0’ is ambiguous
    Note: there are several potential instances:
      instance Fractional Double -- Defined in ‘GHC.Float’
      instance Fractional Float -- Defined in ‘GHC.Float’
      instance Integral a => Fractional (GHC.Real.Ratio a)
        -- Defined in ‘GHC.Real’
    In the second argument of ‘($)’, namely
      ‘(fromIntegral $ VU.length vec) / (fromIntegral n)’
    In the expression:
      ceiling $ (fromIntegral $ VU.length vec) / (fromIntegral n)
    In an equation for ‘lenmax’:
        lenmax
          = ceiling $ (fromIntegral $ VU.length vec) / (fromIntegral n)
Failed to install ConstraintKinds-1.1.0.0
cabal: Error: some packages failed to install:
ConstraintKinds-1.1.0.0 failed during the building phase. The exception was:
ExitFailure 1
```

**Comment by wildman1234** [comment]
This makes it incompatible with HLearn-algebra (why latest version only allows 1.1.0.0 I guess). 
In which version of GHC does 1.1.0.0 compile?

**Comment by slon1024** [comment]
Yeap, I also can't install Hlearn-\*. I try fixed manually **ConstraintKinds-1.1.0.0** by  removed three files: **Partitionable.hs**, **Filterable.hs**, **Traversable.hs** (how mentioned @mikeizbicki above about a new version **ConstraintKinds-1.3.0**). After this I installed **ConstraintKinds-1.1.0.0**:
```
➜  cabal info ConstraintKinds
* ConstraintKinds  (library)
    Synopsis:      Repackages standard type classes with the ConstraintKinds
                   extension
    Versions available: 0.0.1, 0.0.1.1, 0.0.1.2, 0.0.1.3, 0.0.2.0, 1.1.0.0,
                        1.3.0
    Versions installed: 1.1.0.0
...
```

But this approach failed in the next step. For example **HAlgebra** need **Partitionable**:
```
Building HLearn-algebra-1.1.0.1...
Preprocessing library HLearn-algebra-1.1.0.1...
[ 1 of 15] Compiling HLearn.Algebra.Types.Nat1 ( src/HLearn/Algebra/Types/Nat1.hs, dist/build/HLearn/Algebra/Types/Nat1.o )
[ 2 of 15] Compiling HLearn.Algebra.Structures.Triangles ( src/HLearn/Algebra/Structures/Triangles.hs, dist/build/HLearn/Algebra/Structures/Triangles.o )
[ 3 of 15] Compiling HLearn.Algebra.Structures.Groups ( src/HLearn/Algebra/Structures/Groups.hs, dist/build/HLearn/Algebra/Structures/Groups.o )
[ 4 of 15] Compiling HLearn.Algebra.Structures.Modules ( src/HLearn/Algebra/Structures/Modules.hs, dist/build/HLearn/Algebra/Structures/Modules.o )
[ 5 of 15] Compiling HLearn.Algebra.Structures.MetricSpace ( src/HLearn/Algebra/Structures/MetricSpace.hs, dist/build/HLearn/Algebra/Structures/MetricSpace.o )
[ 6 of 15] Compiling HLearn.Algebra.Types.HList ( src/HLearn/Algebra/Types/HList.hs, dist/build/HLearn/Algebra/Types/HList.o )

src/HLearn/Algebra/Types/HList.hs:118:5:
    ‘typeOf’ is not a (visible) method of class ‘Typeable’

src/HLearn/Algebra/Types/HList.hs:288:1:
    Kind variable also used as type variable: ‘a’
    In the declaration for type family ‘$’
[ 8 of 15] Compiling HLearn.Algebra.Functions ( src/HLearn/Algebra/Functions.hs, dist/build/HLearn/Algebra/Functions.o )

src/HLearn/Algebra/Functions.hs:47:7:
    Not in scope: type constructor or class ‘CK.Partitionable’

src/HLearn/Algebra/Functions.hs:48:7:
    Not in scope:
      type constructor or class ‘CK.PartitionableConstraint’
```

@mikeizbicki any advice/tips how to install **HLearn-classification** right now?

**Comment by mikeizbicki** [comment]
The HLearn-* packages on hackage are currently deprecated and unsupported.  These packages had a number of problems that made them less useful in practice.  The master branch of the HLearn repo fixes these problems and adds lots of new features.  Unfortunately, it's not super well documented at this point.  But that's where I'm spending my time, and I'm not sure how long until it's ready for a real release.

If you *really* want HLearn-classification installed for some reason, you can install it in a sandbox by adjusting all the inequality constraints to equalities in the cabal file (i.e. change `package >= 1.1.1` to `package == 1.1.1`).  I know the package installed and worked under those conditions.
